src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,485 @@
+            });
+        }).then([this, fs_shards_pool_size, fs_shard_id] {
+            // Initialize _cluser_allocator
+            std::deque<cluster_id_t> free_clusters;
+            for (auto cid : boost::irange(_available_clusters.beg, _available_clusters.end)) {
+                if (_taken_clusters.count(cid) == 0) {
+                    free_clusters.emplace_back(cid);
+                }
+            }
+            _metadata_log._cluster_allocator = cluster_allocator(std::move(_taken_clusters), std::move(free_clusters));
+
+            // Reset _inode_allocator
+            std::optional<inode_t> max_inode_no;
+            if (not _metadata_log._inodes.empty()) {
+                max_inode_no = (--_metadata_log._inodes.end())->first;
+            }
+            _metadata_log._inode_allocator = shard_inode_allocator(fs_shards_pool_size, fs_shard_id, max_inode_no);
+
+            // TODO: what about orphaned inodes: maybe they are remnants of unlinked files and we need to delete them, or maybe not?
+        });
+    }
+
+    future<> bootstrap_cluster(cluster_id_t curr_cluster) {
+        disk_offset_t curr_cluster_disk_offset = cluster_id_to_offset(curr_cluster, _metadata_log._cluster_size);
+        return _metadata_log._device.read(curr_cluster_disk_offset, _curr_cluster_data.get_write(), _metadata_log._cluster_size).then([this](size_t bytes_read) {
+            if (bytes_read != _metadata_log._cluster_size) {
+                return make_exception_future(std::runtime_error("Failed to read whole cluster of the metadata log"));
+            }
+
+            _curr_cluster = data_reader(_curr_cluster_data.get(), _metadata_log._cluster_size);
+            return bootstrap_read_cluster();
+        });
+    }
+
+    static auto invalid_entry_exception() {
+        return make_exception_future<>(std::runtime_error("Invalid metadata log entry"));
+    }
+
+    future<> bootstrap_read_cluster() {
+        // Process cluster: the data layout format is:
+        // | checkpoint1 | data1... | checkpoint2 | data2... | ... |
+        return do_with(false, [this](bool& whole_log_ended) {
+            return do_until([this, &whole_log_ended] { return whole_log_ended or _next_cluster.has_value(); }, [this, &whole_log_ended] {
+                if (not read_and_check_checkpoint()) {
+                    whole_log_ended = true;
+                    return now();
+                }
+
+                return bootstrap_checkpointed_data();
+            });
+        });
+    }
+
+    // Returns whether reading and checking was successful
+    bool read_and_check_checkpoint() {
+        ondisk_type entry_type;
+        ondisk_checkpoint checkpoint;
+        if (not _curr_cluster.read_entry(entry_type) or entry_type != CHECKPOINT or not _curr_cluster.read_entry(checkpoint)) {
+            return false;
+        }
+
+        boost::crc_32_type crc;
+        if (not _curr_cluster.process_crc_without_reading(crc, checkpoint.checkpointed_data_length)) {
```

Comment:
Probably we would like to have some debug logs to find out easily that something went wrong during bootstrapping.

Reply:
Indeed. Example:
```cpp
#include <seastar/util/log.hh>
logger mlogger("fs_metadata");
...
mlogger.warn("Something went wrong: {}", something);
```